File: backend/src/app/configbackup/store.py
"""Persist + prune agent-pushed config.xml versions (encrypted at rest).

The agent sends ``{sha256, size, content_gz_b64}`` only when the file changed;
``record_config_backup`` additionally dedupes against the newest stored version
so an agent restart (which re-pushes its baseline) never creates a duplicate
row. Content is stored as a Fernet token — config.xml carries secrets. The pure
helpers (``decode_payload``, ``unified_config_diff``) carry the logic and are
unit-tested; the DB functions are thin wrappers verified live.
"""

from __future__ import annotations

import asyncio
import base64
import difflib
import hashlib
import zlib

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.crypto.secrets import decrypt, encrypt
from app.db.models import ConfigBackup
from app.logs.store import surplus_ids

KEEP_PER_INSTANCE = 30
# Plaintext cap. MEDIUMBLOB holds 16 MB; the Fernet token of 8 MB stays under that.
MAX_BYTES = 8_000_000
DIFF_MAX_LINES = 4000
# Refuse to diff versions beyond this many lines (SequenceMatcher DoS guard).
DIFF_MAX_INPUT_LINES = 150_000


def _gunzip_capped(data: bytes, max_bytes: int) -> bytes | None:
    """Gunzip with an output cap so a tiny compressed bomb can't balloon memory."""
    d = zlib.decompressobj(wbits=31)  # 31 = gzip container
    try:
        out = d.decompress(data, max_bytes + 1)
    except zlib.error:
        return None
    if len(out) > max_bytes or not d.eof:
        return None
    return out


def decode_payload(payload: object, max_bytes: int = MAX_BYTES) -> tuple[str, str] | None:
    """Validate an agent config push into ``(sha256, xml_text)``.

    Rejects malformed payloads, over-cap content, and content whose sha256
    doesn't match the agent's claim (truncated/corrupt transfer)."""
    if not isinstance(payload, dict):
        return None
    b64 = payload.get("content_gz_b64")
    claimed = str(payload.get("sha256") or "")
    if not b64 or not claimed:
        return None
    try:
        gz = base64.b64decode(b64, validate=True)
    except (ValueError, TypeError):
        return None
    raw = _gunzip_capped(gz, max_bytes)
    if raw is None:
        return None
    if hashlib.sha256(raw).hexdigest() != claimed:
        return None
    return claimed, raw.decode("utf-8", errors="replace")


def unified_config_diff(
    a: str,
    b: str,
    from_label: str,
    to_label: str,
    max_lines: int = DIFF_MAX_LINES,
    max_input_lines: int = DIFF_MAX_INPUT_LINES,
) -> tuple[str, bool]:
    """Unified diff between two config texts, capped at ``max_lines`` lines.

    CPU-bound (SequenceMatcher is ~O(n*m) on line counts) — callers in async
    context must run it via ``asyncio.to_thread``. Oversized inputs are refused
    outright so a pathological version pair can't burn minutes of CPU."""
    a_lines = a.splitlines()
    b_lines = b.splitlines()
    if max(len(a_lines), len(b_lines)) > max_input_lines:
        return "(versions too large to diff — download both and compare locally)", True
    out: list[str] = []
    for line in difflib.unified_diff(
        a_lines, b_lines, fromfile=from_label, tofile=to_label, lineterm=""
    ):
        if len(out) >= max_lines:
            return "\n".join(out), True
        out.append(line)
    return "\n".join(out), False


async def _newest_first_ids(session: AsyncSession, instance_id: int) -> list[int]:
    rows = await session.execute(
        select(ConfigBackup.id)
        .where(ConfigBackup.instance_id == instance_id)
        .order_by(ConfigBackup.collected_at.desc(), ConfigBackup.id.desc())
    )
    return list(rows.scalars().all())


async def _latest_sha(session: AsyncSession, instance_id: int) -> str | None:
    row = await session.execute(
        select(ConfigBackup.sha256)
        .where(ConfigBackup.instance_id == instance_id)
        .order_by(ConfigBackup.collected_at.desc(), ConfigBackup.id.desc())
        .limit(1)
    )
    return row.scalars().first()


async def record_config_backup(
    session: AsyncSession, instance_id: int, payload: object, source: str = "agent"
) -> bool:
    """Store one pushed config version; dedupe against the newest stored sha.

    Returns True when a new version row was created."""
    decoded = decode_payload(payload)
    if decoded is None:
        return False
    sha, text = decoded
    if await _latest_sha(session, instance_id) == sha:
        return False
    # encrypt() on a multi-MB config is CPU-bound — keep it off the event loop
    # (this runs inside the agent-ingest path, which serves every agent).
    content_enc = await asyncio.to_thread(encrypt, text)
    session.add(
        ConfigBackup(
            instance_id=instance_id,
            sha256=sha,
            bytes=len(text.encode()),
            source=source,
            content_enc=content_enc,
        )
    )
    await session.flush()
    extra = surplus_ids(await _newest_first_ids(session, instance_id), keep=KEEP_PER_INSTANCE)
    if extra:
        await session.execute(delete(ConfigBackup).where(ConfigBackup.id.in_(extra)))
    return True


async def list_config_backups(session: AsyncSession, instance_id: int) -> list[ConfigBackup]:
    rows = await session.execute(
        select(
            ConfigBackup.id,
            ConfigBackup.collected_at,
            ConfigBackup.sha256,
            ConfigBackup.bytes,
            ConfigBackup.source,
        )
        .where(ConfigBackup.instance_id == instance_id)
        .order_by(ConfigBackup.collected_at.desc(), ConfigBackup.id.desc())
    )
    return list(rows.all())


async def get_config_backup(
    session: AsyncSession, instance_id: int, backup_id: int
) -> ConfigBackup | None:
    """A single stored version, scoped to the instance (cross-instance ids miss)."""
    row = await session.execute(
        select(ConfigBackup).where(
            ConfigBackup.id == backup_id, ConfigBackup.instance_id == instance_id
        )
    )
    return row.scalars().first()


def config_text(row: ConfigBackup) -> str:
    """Decrypt a stored version back to XML text (server-side only)."""
    return decrypt(row.content_enc)

File: orbit/lib/orbit/config_backup/store.ex
defmodule Orbit.ConfigBackup.Store do
  @moduledoc """
  Persist agent-pushed config-backup snapshots, Fernet-encrypted at rest, keeping
  the newest `keep_per_instance/0` per instance. Mirror of
  backend/src/app/configbackup/store.py.

  `decode_payload/1` is pure and unit-tested; the DB path (encrypt + dedup +
  prune) is verified live (the test DB is schemaless under the Alembic monopoly).
  """

  import Ecto.Query

  alias Orbit.ConfigBackup.Backup
  alias Orbit.Repo

  @keep_per_instance 30
  @max_bytes 8_000_000

  @doc "Newest-versions-per-instance retention window."
  def keep_per_instance, do: @keep_per_instance

  @doc """
  Validate an agent config push into `{:ok, {sha256, xml_text}}`. Rejects a
  non-dict, missing/blank fields, bad base64, over-cap or corrupt gzip, and
  content whose sha256 doesn't match the agent's claim (truncated transfer).
  """
  @spec decode_payload(term()) :: {:ok, {String.t(), binary()}} | :error
  def decode_payload(%{"content_gz_b64" => b64, "sha256" => claimed})
      when is_binary(b64) and is_binary(claimed) and b64 != "" and claimed != "" do
    with {:ok, gz} <- Base.decode64(b64),
         {:ok, raw} <- gunzip_capped(gz),
         ^claimed <- Base.encode16(:crypto.hash(:sha256, raw), case: :lower) do
      {:ok, {claimed, raw}}
    else
      _ -> :error
    end
  end

  def decode_payload(_), do: :error

  # Post-gunzip length cap. A streaming cap (reject before full inflate) is a
  # documented hardening seam; the agent is signed + trusted and this runs off
  # the hub loop in a Task.
  defp gunzip_capped(gz) do
    raw = :zlib.gunzip(gz)
    if byte_size(raw) > @max_bytes, do: :error, else: {:ok, raw}
  rescue
    _ -> :error
  catch
    _, _ -> :error
  end

  @doc """
  Store one pushed config version, deduped against the newest stored sha.
  Returns true when a new version row was created.
  """
  @spec record(integer(), term(), String.t()) :: boolean()
  def record(instance_id, payload, source \\ "agent") do
    with {:ok, {sha, text}} <- decode_payload(payload),
         false <- latest_sha(instance_id) == sha do
      Repo.insert!(%Backup{
        instance_id: instance_id,
        sha256: sha,
        bytes: byte_size(text),
        source: source,
        content_enc: Orbit.Crypto.encrypt(text),
        collected_at: DateTime.utc_now() |> DateTime.truncate(:second)
      })

      prune(instance_id)
      true
    else
      _ -> false
    end
  end

  defp prune(instance_id) do
    ids =
      Repo.all(
        from(b in Backup,
          where: b.instance_id == ^instance_id,
          order_by: [desc: b.collected_at, desc: b.id],
          select: b.id
        )
      )

    case Enum.drop(ids, @keep_per_instance) do
      [] -> :ok
      extra -> Repo.delete_all(from(b in Backup, where: b.id in ^extra))
    end
  end

  @doc "Newest stored sha for an instance, or nil."
  @spec latest_sha(integer()) :: String.t() | nil
  def latest_sha(instance_id) do
    Repo.one(
      from(b in Backup,
        where: b.instance_id == ^instance_id,
        order_by: [desc: b.collected_at, desc: b.id],
        limit: 1,
        select: b.sha256
      )
    )
  end

  @doc "Version metadata for an instance, newest-first (no content)."
  @spec list(integer()) :: [map()]
  def list(instance_id) do
    Repo.all(
      from(b in Backup,
        where: b.instance_id == ^instance_id,
        order_by: [desc: b.collected_at, desc: b.id],
        select: %{
          id: b.id,
          collected_at: b.collected_at,
          sha256: b.sha256,
          bytes: b.bytes,
          source: b.source
        }
      )
    )
  end

  @diff_max_lines 4000
  @diff_max_input_lines 150_000

  @doc """
  Line diff of one version against the chronologically previous one for the same
  instance. `{:ok, text, truncated?}` when both exist, `:no_previous` when the
  version is the oldest, `:error` when a version is missing/undecryptable.

  CPU-bound (Myers is ~O(n*m) on line counts) — callers run it off the loop.
  """
  @spec diff_against_previous(integer(), integer()) ::
          {:ok, String.t(), boolean()} | :no_previous | :error
  def diff_against_previous(instance_id, backup_id) do
    prev_id =
      Repo.one(
        from(b in Backup,
          where: b.instance_id == ^instance_id and b.id < ^backup_id,
          order_by: [desc: b.id],
          limit: 1,
          select: b.id
        )
      )

    cond do
      is_nil(prev_id) -> :no_previous
      true -> do_diff(get_content(instance_id, prev_id), get_content(instance_id, backup_id))
    end
  end

  @doc "Diff two arbitrary stored versions (inline viewer). Same bounds as above."
  def diff_between(instance_id, id_a, id_b) do
    do_diff(get_content(instance_id, id_a), get_content(instance_id, id_b))
  end

  defp do_diff(a, b) when is_binary(a) and is_binary(b) do
    a_lines = String.split(a, "\n")
    b_lines = String.split(b, "\n")

    if max(length(a_lines), length(b_lines)) > @diff_max_input_lines do
      {:ok, "(versions too large to diff — download both and compare locally)", true}
    else
      lines =
        a_lines
        |> List.myers_difference(b_lines)
        |> Enum.flat_map(&diff_chunk/1)

      {kept, truncated} = Enum.split(lines, @diff_max_lines)
      {:ok, Enum.join(kept, "\n"), truncated != []}
    end
  end

  defp do_diff(_, _), do: :error

  defp diff_chunk({:eq, _lines}), do: []
  defp diff_chunk({:del, lines}), do: Enum.map(lines, &("-" <> &1))
  defp diff_chunk({:ins, lines}), do: Enum.map(lines, &("+" <> &1))

  @doc "Decrypted XML for one version (scoped to the instance), or nil."
  @spec get_content(integer(), integer()) :: String.t() | nil
  def get_content(instance_id, id) do
    enc =
      Repo.one(
        from(b in Backup,
          where: b.instance_id == ^instance_id and b.id == ^id,
          select: b.content_enc
        )
      )

    case enc && Orbit.Crypto.decrypt(enc) do
      {:ok, text} -> text
      _ -> nil
    end
  end
end
